File: paleo_workbench/ui/layout_presets.py
"""Named workstation layout presets (V3 Light design-qa P3).

Pure data + visibility matrices — no Qt widgets. ``WorkstationFrame`` (workstation shell) applies
these through ``QMainWindow`` dock show/hide and optional ``saveState``
snapshots. Existing ``WorkspacePreset`` values (MAP_AUTHORING, etc.) stay
untouched; workstation ids live beside them via :func:`register_with_dock_manager`.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable


#: Deprecated document-tab keys (Task 2 removes shell usages, then delete).
TAB_COMPOSITE = "composite"
TAB_JOINT = "joint"


@dataclass(frozen=True)
class DockVisibilityMatrix:
    """Visibility flags for the top-level workstation dock set.

    Composite docks only matter when the composite document is active; the
    shell still applies the flags so switching back to 编图 restores them.
    """

    nav: bool = True
    inspector: bool = True
    process: bool = False
    tasks: bool = False
    composite_layer: bool = True
    composite_input: bool = False
    composite_linked: bool = False
    explorer_expanded: bool = True
    well: bool = False
    seismic: bool = False
    hub: bool = False


@dataclass(frozen=True)
class WorkstationLayoutPreset:
    id: str
    label: str
    description: str
    visibility: DockVisibilityMatrix
    #: When True, float every currently-visible shell dock (low-risk affordance).
    float_visible: bool = False


#: Canonical registry — order is menu order.
WORKSTATION_LAYOUT_PRESETS: tuple[WorkstationLayoutPreset, ...] = (
    WorkstationLayoutPreset(
        id="composite_default",
        label="默认编图",
        description="Variant C full-bleed map; only the layer dock open among composite panels.",
        visibility=DockVisibilityMatrix(
            nav=True,
            inspector=True,
            process=False,
            tasks=False,
            composite_layer=True,
            composite_input=False,
            composite_linked=False,
            explorer_expanded=True,
            well=False,
            seismic=False,
            hub=False,
        ),
    ),
    WorkstationLayoutPreset(
        id="interpretation",
        label="解释工作区",
        description="Explorer + inspector + tasks open for linked interpretation.",
        visibility=DockVisibilityMatrix(
            nav=True,
            inspector=True,
            process=True,
            tasks=True,
            composite_layer=True,
            composite_input=False,
            composite_linked=False,
            explorer_expanded=True,
            well=True,
            seismic=True,
            hub=False,
        ),
    ),
)

#: Alias used by the 面板 menu "恢复默认布局" action.
RESET_LAYOUT_PRESET_ID = "composite_default"


def list_presets() -> tuple[WorkstationLayoutPreset, ...]:
    return WORKSTATION_LAYOUT_PRESETS


def get_preset(preset_id: str) -> WorkstationLayoutPreset | None:
    for preset in WORKSTATION_LAYOUT_PRESETS:
        if preset.id == preset_id:
            return preset
    return None


def preset_labels() -> tuple[tuple[str, str], ...]:
    """``(id, label)`` pairs for UI menus."""
    return tuple((p.id, p.label) for p in WORKSTATION_LAYOUT_PRESETS)


def visibility_dict(matrix: DockVisibilityMatrix) -> dict[str, bool]:
    """Flat dock-key → visible map (stable keys for tests / persistence notes)."""
    return {
        "nav": matrix.nav,
        "inspector": matrix.inspector,
        "process": matrix.process,
        "tasks": matrix.tasks,
        "composite_layer": matrix.composite_layer,
        "composite_input": matrix.composite_input,
        "composite_linked": matrix.composite_linked,
        "explorer_expanded": matrix.explorer_expanded,
        "well": matrix.well,
        "seismic": matrix.seismic,
        "hub": matrix.hub,
    }


def register_with_dock_manager(dock_manager) -> None:
    """Register workstation panel titles into the shared DockManager vocabulary.

    Does not replace existing MAP_AUTHORING / WELL_LOG presets; only seeds
    panel ids used by the V3 shell and FloatController title lookup.
    """
    panels: Iterable[tuple[str, str, str]] = (
        ("workstation:explorer", "资源管理器", "left"),
        ("workstation:inspector", "检查器", "right"),
        ("workstation:process", "Agent", "bottom"),
        ("workstation:tasks", "任务中心", "bottom"),
        ("workstation:composite_layer", "图层管理", "right"),
        ("workstation:composite_input", "输入与结果", "left"),
        ("workstation:composite_linked", "联动视图", "bottom"),
        ("workstation:well", "测井轨道", "bottom"),
        ("workstation:seismic", "地震剖面", "bottom"),
        ("workstation:hub", "功能页", "right"),
    )
    for panel_id, title, area in panels:
        dock_manager.register_panel(panel_id, title, area=area)

File: libs/ui_shell/src/layout_presets.cpp
#include "pwb/ui_shell/layout_presets.hpp"

#include "pwb/ui_shell/dock_manager.hpp"

namespace pwb::ui_shell {

const std::vector<WorkstationLayoutPreset>& workstation_layout_presets() {
    static const std::vector<WorkstationLayoutPreset> presets = {
        {"composite_default",
         "编图 · 默认",
         "Variant C full-bleed map; only the layer dock open among composite "
         "panels.",
         DockVisibilityMatrix{.nav = true,
                              .inspector = true,
                              .composite_layer = true,
                              .explorer_expanded = true}},
        {"well_interpretation",
         "测井解释",
         "Explorer + inspector + well tracks + layer dock; no "
         "agent/tasks/seismic.",
         DockVisibilityMatrix{.nav = true,
                              .inspector = true,
                              .composite_layer = true,
                              .explorer_expanded = true,
                              .well = true}},
        {"seismic_interpretation",
         "地震解释",
         "Explorer + inspector + seismic sections + layer dock.",
         DockVisibilityMatrix{.nav = true,
                              .inspector = true,
                              .composite_layer = true,
                              .explorer_expanded = true,
                              .seismic = true}},
        {"well_seismic_joint",
         "井震联合",
         "Well tracks + seismic sections side by side with the linked view.",
         DockVisibilityMatrix{.nav = true,
                              .inspector = true,
                              .composite_layer = false,
                              .composite_linked = true,
                              .explorer_expanded = true,
                              .well = true,
                              .seismic = true}},
        {"integrated",
         "综合",
         "Everything on: agent, tasks, wells, seismic and the linked view.",
         DockVisibilityMatrix{.nav = true,
                              .inspector = true,
                              .agent = true,
                              .tasks = true,
                              .composite_layer = true,
                              .composite_input = true,
                              .composite_linked = true,
                              .explorer_expanded = true,
                              .well = true,
                              .seismic = true}},
        {"review",
         "审核",
         "Inputs and results dock open next to the layer dock for review.",
         DockVisibilityMatrix{.nav = true,
                              .inspector = true,
                              .composite_layer = true,
                              .composite_input = true,
                              .explorer_expanded = true}},
    };
    return presets;
}

const WorkstationLayoutPreset* get_preset(const std::string& preset_id) {
    for (const auto& preset : workstation_layout_presets()) {
        if (preset.id == preset_id) {
            return &preset;
        }
    }
    return nullptr;
}

std::vector<std::pair<std::string, std::string>> preset_labels() {
    std::vector<std::pair<std::string, std::string>> out;
    out.reserve(workstation_layout_presets().size());
    for (const auto& p : workstation_layout_presets()) {
        out.emplace_back(p.id, p.label);
    }
    return out;
}

std::map<std::string, bool> visibility_dict(const DockVisibilityMatrix& m) {
    return {
        {"nav", m.nav},
        {"inspector", m.inspector},
        {"agent", m.agent},
        {"tasks", m.tasks},
        {"logs", m.logs},
        {"console", m.console},
        {"composite_layer", m.composite_layer},
        {"composite_input", m.composite_input},
        {"composite_linked", m.composite_linked},
        {"explorer_expanded", m.explorer_expanded},
        {"well", m.well},
        {"seismic", m.seismic},
        {"hub", m.hub},
        {"mapping_stage", m.mapping_stage},
    };
}

void register_with_dock_manager(DockManager& dock_manager) {
    struct Row {
        const char* id;
        const char* title;
        const char* area;
    };
    static const Row panels[] = {
        {"workstation:explorer", "资源管理器", "left"},
        {"workstation:inspector", "检查器", "right"},
        {"workstation:agent", "Agent", "bottom"},
        {"workstation:tasks", "任务中心", "bottom"},
        {"workstation:logs", "日志", "bottom"},
        {"workstation:console", "控制台", "bottom"},
        {"workstation:composite_layer", "图层管理", "right"},
        {"workstation:composite_input", "输入与结果", "left"},
        {"workstation:composite_linked", "联动视图", "bottom"},
        {"workstation:well", "测井轨道", "bottom"},
        {"workstation:seismic", "地震剖面", "bottom"},
        {"workstation:hub", "功能页", "right"},
    };
    for (const auto& row : panels) {
        dock_manager.register_panel(row.id, row.title, row.area);
    }
}

}  // namespace pwb::ui_shell
